controllers/argocd/server/role.go

```diff
@@ -0,0 +1,411 @@
+package server
+
+import (
+	"reflect"
+
+	"github.com/argoproj-labs/argocd-operator/common"
+	"github.com/argoproj-labs/argocd-operator/pkg/cluster"
+	"github.com/argoproj-labs/argocd-operator/pkg/mutation"
+	"github.com/argoproj-labs/argocd-operator/pkg/permissions"
+	rbacv1 "k8s.io/api/rbac/v1"
+	v1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	amerr "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// reconcileRoles will ensure all ArgoCD Server roles are present
+func (sr *ServerReconciler) reconcileRoles() error {
+	sr.Logger.V(0).Info("reconciling role")
+
+	var reconciliationErrors []error
+
+	// reconcile Roles for managed namespaces
+	if err := sr.reconcileManagedRoles(); err != nil {
+		reconciliationErrors = append(reconciliationErrors, err)
+	}
+
+	// reconcile Roles for source namespaces
+	if err := sr.reconcileSourceRoles(); err != nil {
+		reconciliationErrors = append(reconciliationErrors, err)
+	}
+
+	return amerr.NewAggregate(reconciliationErrors)
+}
+
+// reconcileManagedRoles manages roles in ArgoCD managed namespaces
+func (sr *ServerReconciler) reconcileManagedRoles() error {
+	var reconciliationErrors []error
+
+	for nsName := range sr.ManagedNamespaces {
+
+		roleName := getRoleName(sr.Instance.Name)
+		roleLabels := common.DefaultLabels(roleName, sr.Instance.Name, ServerControllerComponent)
+
+		_, err := cluster.GetNamespace(nsName, sr.Client)
+		if err != nil {
+			if !errors.IsNotFound(err){
+				sr.Logger.Error(err, "reconcileManagedRoles: failed to retrieve namespace", "name", nsName)
+				reconciliationErrors = append(reconciliationErrors, err)
+			}
+			continue
+		}
+		
+		roleRequest := permissions.RoleRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        roleName,
+				Labels:      roleLabels,
+				Annotations: sr.Instance.Annotations,
+			},
+			Client:    sr.Client,
+			Mutations: []mutation.MutateFunc{mutation.ApplyReconcilerMutation},
+		}
+
+		roleRequest.ObjectMeta.Namespace = nsName
```

Comment:
can we move this to where the rest of the fields are being initialized in the roleRequest just above? seems out of place to set this separately here